### PR TITLE
Add `manage_shards` to interactively remove compiler shards

### PR DESCRIPTION
**Pull request by sloede**
This PR is inspired by #206 (and, if merged, supersedes it) and based on its code by @Crghilardi. It adds a `manage_shards` function that allows one to interactively view the currently installed compiler shards and then remove them. This can be helpful if one experiments with a lot of BB builds and ends up with tens of GB of compiled shards that clutter up the Julia depot.

Default usage (without deleting anything):
```julia
julia> using BinaryBuilderBase

julia> manage_shards()
Which compiler shards should be removed (total size on disk: 79.666 GiB)?
[press: d=done, a=all, n=none]
 > [ ] (4.062 GiB) CompilerShard("GCCBootstrap", v"8.1.0", macOS x86_64 {os_version=14.0.0}, Linux x86_64 {libc=musl}, :unpacked)
   [ ] (3.179 GiB) CompilerShard("GCCBootstrap", v"4.8.5", macOS x86_64 {os_version=14.0.0}, Linux x86_64 {libc=musl}, :unpacked)
   [ ] (2.649 GiB) CompilerShard("GCCBootstrap", v"10.2.0", Linux aarch64 {libc=glibc}, Linux x86_64 {libc=musl}, :unpacked)
   [ ] (2.635 GiB) CompilerShard("GCCBootstrap", v"10.2.0", Linux x86_64 {libc=glibc}, Linux x86_64 {libc=musl}, :unpacked)
   [ ] (2.479 GiB) CompilerShard("GCCBootstrap", v"10.2.0", Linux x86_64 {libc=musl}, Linux x86_64 {libc=musl}, :unpacked)
   [ ] (2.470 GiB) CompilerShard("GCCBootstrap", v"10.2.0", Linux i686 {libc=glibc}, Linux x86_64 {libc=musl}, :unpacked)
   [ ] (2.426 GiB) CompilerShard("GCCBootstrap", v"10.2.0", FreeBSD x86_64 {os_version=12.2.0}, Linux x86_64 {libc=musl}, :unpacked)
   [ ] (2.414 GiB) CompilerShard("GCCBootstrap", v"10.2.0", Linux aarch64 {libc=musl}, Linux x86_64 {libc=musl}, :unpacked)
   [ ] (2.359 GiB) CompilerShard("GCCBootstrap", v"10.2.0", Linux i686 {libc=musl}, Linux x86_64 {libc=musl}, :unpacked)
v  [ ] (2.307 GiB) CompilerShard("GCCBootstrap", v"10.2.0", Linux armv7l {call_abi=eabihf, libc=glibc}, Linux x86_64 {libc=musl}, :unpacked)
Removed 0 compiler shards.
```

Sorted by size on disk and deleting 2 shards:
```julia
julia> manage_shards(sort_by=:size, rev=true)
Which compiler shards should be removed (total size on disk: 79.666 GiB)?
[press: d=done, a=all, n=none]
   [X] (4.609 GiB) BinaryBuilderBase.CompilerShard("GCCBootstrap", v"10.2.0", macOS x86_64 {os_version=14.0.0}, Linux x86_64 {libc=musl}, :unpacked)
   [ ] (4.062 GiB) BinaryBuilderBase.CompilerShard("GCCBootstrap", v"8.1.0", macOS x86_64 {os_version=14.0.0}, Linux x86_64 {libc=musl}, :unpacked)
   [ ] (3.179 GiB) BinaryBuilderBase.CompilerShard("GCCBootstrap", v"4.8.5", macOS x86_64 {os_version=14.0.0}, Linux x86_64 {libc=musl}, :unpacked)
 > [X] (2.649 GiB) BinaryBuilderBase.CompilerShard("GCCBootstrap", v"10.2.0", Linux aarch64 {libc=glibc}, Linux x86_64 {libc=musl}, :unpacked)
   [ ] (2.635 GiB) BinaryBuilderBase.CompilerShard("GCCBootstrap", v"10.2.0", Linux x86_64 {libc=glibc}, Linux x86_64 {libc=musl}, :unpacked)
   [ ] (2.481 GiB) BinaryBuilderBase.CompilerShard("GCCBootstrap", v"10.2.0", Linux powerpc64le {libc=glibc}, Linux x86_64 {libc=musl}, :unpacked)
   [ ] (2.479 GiB) BinaryBuilderBase.CompilerShard("GCCBootstrap", v"10.2.0", Linux x86_64 {libc=musl}, Linux x86_64 {libc=musl}, :unpacked)
   [ ] (2.470 GiB) BinaryBuilderBase.CompilerShard("GCCBootstrap", v"10.2.0", Linux i686 {libc=glibc}, Linux x86_64 {libc=musl}, :unpacked)
   [ ] (2.426 GiB) BinaryBuilderBase.CompilerShard("GCCBootstrap", v"10.2.0", FreeBSD x86_64 {os_version=12.2.0}, Linux x86_64 {libc=musl}, :unpacked)
v  [ ] (2.414 GiB) BinaryBuilderBase.CompilerShard("GCCBootstrap", v"10.2.0", Linux aarch64 {libc=musl}, Linux x86_64 {libc=musl}, :unpacked)
The following shards have been marked for removal:

  (4.609 GiB) BinaryBuilderBase.CompilerShard("GCCBootstrap", v"10.2.0", macOS x86_64 {os_version=14.0.0}, Linux x86_64 {libc=musl}, :unpacked)
  (2.649 GiB) BinaryBuilderBase.CompilerShard("GCCBootstrap", v"10.2.0", Linux aarch64 {libc=glibc}, Linux x86_64 {libc=musl}, :unpacked)

Are you sure that these should be removed? [Y/n]: y
Removed 2 compiler shards.
```

As stated by Casey, it is unclear if something like this is even wanted in BBB, but I found it hugely useful for myself, and I thought I'd share it here.

### CODE REVIEW

File: src/Rootfs.jl [hunk 1]
```diff
@@ -1100,3 +1107,83 @@ function installed_shards()
...
+    shards = installed_shards()
+
+    # Obtain directory sizes for the artifacts
+    totalsizes = totalsize.(Artifacts.artifact_path.(shard_source_artifact_hash.(shards)))
+
+    # Sort shards and totalsizes
+    if sort_by === :name
+        perm = sortperm(shards, rev=rev)
+    elseif sort_by == :size
+        perm = sortperm(totalsizes, rev=rev)
```
Comment (giordano) [inline src/Rootfs.jl:1130]: My simple rule of thumb is: ***ALWAYS*** start keyword arguments with `;`, and then you don't need to care about the few corner cases where the comma wouldn't work
```suggestion
        perm = sortperm(shards; rev)
    elseif sort_by == :size
        perm = sortperm(totalsizes; rev)
```

File: src/Rootfs.jl [hunk 2]
```diff
@@ -57,6 +59,11 @@ function Base.:(==)(a::CompilerShard, b::CompilerShard)
            a.archive_type == b.archive_type
 end
 
+function Base.isless(a::CompilerShard, b::CompilerShard)
+    return isless((a.name, a.version, string(a.target), a.host, a.archive_type),
+                  (b.name, b.version, string(b.target), b.host, b.archive_type))
+end
```
Comment (giordano) [inline src/Rootfs.jl:65]: I don't like defining `isless` for `CompilerShard`s, this looks very arbitrary, but I presume you can define a custom function to be passed to the sorting functions when needed?

File: src/Rootfs.jl [hunk 3]
```diff
@@ -1100,3 +1107,83 @@ function installed_shards()
...
+    menuitems = similar(shards, String)
+    for i in 1:length(shards)
+        menuitems[i] = @sprintf("(%5.3f GiB) %s", totalsizes[i] / 2^30, shards[i])
+    end
+
+    # Prompt user
+    ts = @sprintf("%5.3f", sum(totalsizes) / 2^30)
+    manage_shards_menu = TerminalMenus.request(
+        "Which compiler shards should be removed (total size on disk: $ts GiB)?",
+        TerminalMenus.MultiSelectMenu(menuitems, pagesize = 10; charset=:ascii)
```
Comment (giordano) [inline src/Rootfs.jl:1147]:
```suggestion
        TerminalMenus.MultiSelectMenu(menuitems; pagesize = 10, charset=:ascii)
```

File: src/Rootfs.jl [hunk 4]
```diff
@@ -1100,3 +1107,83 @@ function installed_shards()
...
+        Pkg.remove_artifact.(shard_source_artifact_hash.(shards[Int.(manage_shards_menu)]))
+        println("Removed ", length(manage_shards_menu), " compiler shards.")
+    else
+        println("Removed 0 compiler shards.")
+    end
+end
+
+# Return total size on a disk of a file or directory
+function totalsize(path)
+  if isfile(path)
```
Comment (giordano) [inline src/Rootfs.jl:1175]: Uhm, `isfile` isn't complementary to `isdir`, for example
```julia
julia> isfile("/dev/stdout")
false

julia> isdir("/dev/stdout")
false
```
If I understand correctly here you're assuming that if something `!isfile` then `isdir`?

File: src/Rootfs.jl [hunk 5]
```diff
@@ -1100,3 +1107,83 @@ function installed_shards()
...
+        perm = sortperm(totalsizes, rev=rev)
+    else
+      error("unsupported sort value: :$sort_by (allowed: :name, :size)")
+    end
+    shards .= shards[perm]
+    totalsizes .= totalsizes[perm]
+
+    # Build menu items
+    menuitems = similar(shards, String)
+    for i in 1:length(shards)
```
Comment (giordano) [inline src/Rootfs.jl:1139]:
```suggestion
    for i in eachindex(shards, totalsizes)
```

File: src/Rootfs.jl (hunk 1 above)
Comment (sloede) [reply src/Rootfs.jl]: Thanks, makes total sense!

File: src/Rootfs.jl (hunk 4 above)
Comment (sloede) [reply src/Rootfs.jl]: You're right. I've changed it to `!isdir` in 1f8b1b4, which seems to be closer semantically to what I want (i.e., do not recurse if it does not identify as a directory).

File: src/Rootfs.jl (hunk 2 above)
Comment (sloede) [reply src/Rootfs.jl]: Moved to a local `isless_shards` function where needed in 1ef1746.

File: src/Rootfs.jl [hunk 6]
```diff
@@ -1100,3 +1102,88 @@ function installed_shards()
...
+    shards = installed_shards()
+
+    # Obtain directory sizes for the artifacts
+    totalsizes = totalsize.(Artifacts.artifact_path.(shard_source_artifact_hash.(shards)))
+
+    # Sort shards and totalsizes
+    if sort_by === :name
+        function isless_shards(a::CompilerShard, b::CompilerShard)
+            return isless((a.name, a.version, string(a.target), a.host, a.archive_type),
+                          (b.name, b.version, string(b.target), b.host, b.archive_type))
```
Comment (giordano) [inline src/Rootfs.jl:1125]: I'm getting
```
julia> BinaryBuilderBase.manage_shards()
ERROR: MethodError: no method matching isless(::Platform, ::Platform)

Closest candidates are:
  isless(::Any, ::Missing)
   @ Base missing.jl:88
  isless(::Missing, ::Any)
   @ Base missing.jl:87

Stacktrace:
  [1] isless(t1::Tuple{Platform, Symbol}, t2::Tuple{Platform, Symbol})
    @ Base ./tuple.jl:541
  [2] isless
    @ ./tuple.jl:541 [inlined]
  [3] isless(t1::Tuple{VersionNumber, String, Platform, Symbol}, t2::Tuple{VersionNumber, String, Platform, Symbol})
    @ Base ./tuple.jl:541
  [4] isless(t1::Tuple{String, VersionNumber, String, Platform, Symbol}, t2::Tuple{String, VersionNumber, String, Platform, Symbol})
    @ Base ./tuple.jl:541
  [5] (::BinaryBuilderBase.var"#isless_shards#317")(a::CompilerShard, b::CompilerShard)
    @ BinaryBuilderBase ~/.julia/dev/BinaryBuilderBase/src/Rootfs.jl:1124
```

Comment (sloede) [reply src/Rootfs.jl:1125]: Im afk, but what happens if you change `a.host` to `string(a.host)`?

**Pull request merged**